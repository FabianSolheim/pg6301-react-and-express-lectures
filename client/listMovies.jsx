import React, { useState } from "react";
import { useLoading } from "./useLoading";

function MovieCard({ movie: { title, plot, poster } }) {
  return (
    <>
      <h3>{title}</h3>
      {poster && <img src={poster} width={100} alt={"Movie poster"} />}
      <div>{plot}</div>
    </>
  );
}

export function ListMovies({ movieApi }) {
  const [country, setCountry] = useState();
  const [countryInput, setCountryInput] = useState("");
  const { loading, error, data } = useLoading(
    async () => movieApi.listMovies(country),
    [country]
  );

  if (loading) {
    return <div className="loading-indicator">Loading...</div>;
  }
  if (error) {
    return (
      <div className="error-message">
        <h1>Error</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      <h1>Movies in the database</h1>

      <div>
        <label>Search by country</label>
        <input
          value={countryInput}
          onChange={(e) => setCountryInput(e.target.value)}
        />
        <button onClick={() => setCountry(countryInput)}>Search</button>
      </div>

      {data.map((movie) => (
        <MovieCard key={movie.title} movie={movie} />
      ))}
    </div>
  );
}
