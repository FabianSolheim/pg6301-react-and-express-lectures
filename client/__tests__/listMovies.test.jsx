import { ListMovies } from "../listMovies";
import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";

const movies = [
  {
    title: "My movie",
    plot: "Something happens",
    poster: "https://example.com/poster1.jpg",
  },
  {
    title: "My other movie",
    plot: "Nothing happens",
    poster: "https://example.com/void1.jpg",
  },
];

async function renderListMovies(listMovies) {
  const element = document.createElement("div");
  await act(async () =>
    ReactDOM.render(<ListMovies movieApi={{ listMovies }} />, element)
  );
  return element;
}

describe("list movies", () => {
  it("shows loading screen", async () => {
    const element = await renderListMovies(() => new Promise(() => {}));
    expect(element.querySelector(".loading-indicator")).not.toBeNull();
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("shows movie list", async () => {
    const element = await renderListMovies(
      () => new Promise((resolve) => resolve(movies))
    );
    expect(element.querySelector("h3").innerHTML).toEqual(movies[0].title);
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("shows error message", async () => {
    const element = await renderListMovies(
      () =>
        new Promise((resolve, reject) => {
          reject(new Error("Failed to fetch"));
        })
    );
    expect(element.querySelector(".error-message").innerHTML).toContain(
      "Failed to fetch"
    );
    expect(element.innerHTML).toMatchSnapshot();
  });
});
