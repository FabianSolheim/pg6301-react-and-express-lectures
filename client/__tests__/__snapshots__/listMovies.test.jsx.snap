// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list movies shows error message 1`] = `"<div class=\\"error-message\\"><h1>Error</h1><div>Error: Failed to fetch</div></div>"`;

exports[`list movies shows loading screen 1`] = `"<div class=\\"loading-indicator\\">Loading...</div>"`;

exports[`list movies shows movie list 1`] = `"<div><h1>Movies in the database</h1><h3>My movie</h3><img src=\\"https://example.com/poster1.jpg\\" width=\\"100\\" alt=\\"Movie poster\\"><div>Something happens</div><h3>My other movie</h3><img src=\\"https://example.com/void1.jpg\\" width=\\"100\\" alt=\\"Movie poster\\"><div>Nothing happens</div></div>"`;
